### PR TITLE
Add pre-commit hook for codespell

### LINKED ISSUE #13814: Catch common typos with codespell
### Description

https://github.com/codespell-project/codespell + pre-commit.

### Additional context

We can configure custom common misspells that are Astropy-specific.

### PR DESCRIPTION
**Pull request by namurphy**
### Description

This PR adds a pre-commit hook for [codespell](https://github.com/codespell-project/codespell).  Instead of searching for words that are not found in a dictionary, it looks for words that are in a dictionary of common misspellings. Hence, it gives very few false positives.

It will provide error messages like: 

```
astropy/wcs/utils.py:947: transfomations ==> transformations
```

The codespell configuration in `pyproject.toml` is the place to list false positives and files/directories not to check.  

An additional possibility would be to create a `tox` environment for codespell (discussed in #13814), which is the approach that we took for PlasmaPy.  

Closes #13814.  This will need to be merged after #13982.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label. Codestyle issues can be fixed by the [bot](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit).
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Review by nstarman — APPROVED** [review]
LGTM, thanks @namurphy!

TO MAINTAINERS: DO NOT MERGE UNTIL GREAT BLACKENING & https://github.com/astropy/astropy/pull/13982 ARE DONE.

**Comment by namurphy** [comment]
I've been having some issues with the pre-commit hook for codespell not recognizing the configuration in either `pyproject.toml` or `setup.cfg` in another repo, which appears to be happening here too.  I'll need to look into that later.  The backup option would be to create a tox environment and add it to GitHub Actions.

**Comment by namurphy** [comment]
I just specified the config file for codespell in the pre-commit hook, which should do the trick.  Might need to adjust this later on.

**Comment by saimn** [comment]
How long does it take ?

**Comment by namurphy** [comment]
It only takes a few seconds to run codespell on the whole astropy repo.  It's a lot faster when using pre-commit since pre-commit only looks at the changed files.

**Comment by nstarman** [comment]
Why change to setup.cfg?

**Comment by namurphy** [comment]
> Why change to setup.cfg?

I was getting an error about duplicate keys in `pyproject.toml` when trying to use it as the config file when running codespell.  I had been hoping to use `pyproject.toml` for this, though...in particular given the community-wide shifting towards `pyproject.toml` instead of `setup.cfg`.

File: .pre-commit-config.yaml
Comment (WilliamJamieson) [inline .pre-commit-config.yaml:142]:
```suggestion
  - repo: https://github.com/codespell-project/codespell
    rev: v2.2.2
    hooks:
      - id: codespell
        args: ["--toml=pyproject.toml"]
        additional_dependencies:
          - tomli
```

I was experimenting with setting up `codespell` for one of my other project including using the pre-commit hook (inspired by this and #13982). I found that I needed to modify this so that it would actually use the `toml` settings, these were the necessary changes.

**Review by WilliamJamieson — CHANGES_REQUESTED** [review]

File: .pre-commit-config.yaml
Comment (namurphy) [reply .pre-commit-config.yaml:142]: Thank you!  I'll merge this, and then pose a question about `--write-changes` below.

**Comment by namurphy** [comment]
Right now I'm wondering about the tradeoffs of having `pre-commit` write changes from `codespell`.  

It is rare for `codespell` to yield a false positive; consequently the auto-fixes will *usually* save time and effort.  When there is a false positive, however, an incorrect correction will be applied with a small risk that the change will go unnoticed.

I suspect that it'll work as intended for the vast majority of the time, so my inclination is to try out the `--write-changes` flag, and then revisit it should it end up being more trouble than it's worth.

**Comment by WilliamJamieson** [comment]
> It is rare for `codespell` to yield a false positive; consequently the auto-fixes will *usually* save time and effort.  When there is a false positive, however, an incorrect correction will be applied with a small risk that the change will go unnoticed.

Pre-commit is normally run on the changes staged for commit with the fixes it produces made as unstated changes. Additionally, these changes will be logged by pre-commit (usually an output to the terminal). It is incumbent upon the user to review the changes made by pre-commit before staging those changes. I think this should make it unlikely that incorrect corrections will be applied inadvertently.

File: .pre-commit-config.yaml
Comment (namurphy) [inline .pre-commit-config.yaml:74]: The codespell action with `--write-changes` should precede `black` because it can potentially change line length.

File: .pre-commit-config.yaml
Comment (namurphy) [inline .pre-commit-config.yaml:72]: This line is partially redundant with the `skip` field in the `codespell` configuration in `pyproject.toml`, but appears necessary.  The `pyproject.toml` config is already set to skip `*extern`. However, running `pre-commit run --all-files` did not exclude `extern` and `cextern` from getting processed by codespell.  Weird.

We still want to keep the `skip` field in `pyproject.toml` as-is, since that gives contributors the option to run codespell outside of pre-commit.

File: pyproject.toml
Comment (namurphy) [inline pyproject.toml:272]: Oops!  It appears I added a misspelling in my previous PR to fix misspellings. 👀 😅

**Comment by namurphy** [comment]
Ready for review!

Cross-reference: #14185

File: .pre-commit-config.yaml
Comment (WilliamJamieson) [reply .pre-commit-config.yaml:72]: This is necessary due to the peculiarities of pre-commit. Basically, the way pre-commit works is that it passes each file tracked by git individually to the tool in question rather letting the tool itself iterate over all the files. This is so that pre-commit can pull tricks like running only on files staged for commit rather than having to run on the whole repo.

Often tools are written in a way so that if you explicitly pass a file which it is configured to skip, it will still run on that file. I suspect codespell works that way. Some tools like black have a skip/exclude (force-exclude) that you can set which will never be overridden in this fashion. I wonder if codespell has a similar mechanism?

Comment (nstarman) [reply .pre-commit-config.yaml:72]: Yes. I wish pre-commit had an option to grab the ``exclude`` settings from the tools' configs, like

``exclude: pyproject.toml::codespells::exclude`` so there were only 1 source of configuration. But alas.

File: .pre-commit-config.yaml
Comment (nstarman) [reply .pre-commit-config.yaml:74]: A good point! I need to change my repos now 😆

**Review by nstarman — APPROVED** [review]
Let's get this in before more misspellings are added! Thanks @namurphy

File: pyproject.toml
Comment (WilliamJamieson) [inline pyproject.toml:270]:
```suggestion
skip = "*.dat,*.fits,*.hdr,*.xml,*egg*,*extern/*,.git,.tox,_build,fitshdr.htest_groups.py,venv,_*.c,wcs.h,lin.h,tab.h,spc.h,cython*"
```

I have had a chance to do some digging into `codespell` on the issue of pre-commit not respecting the `skip` arguments, which resulted in adding the `exclude` to the `pre-commit-config.yaml`.

It turns out that `codespell` uses [`fnmatch.fnmatch`](https://docs.python.org/3/library/fnmatch.html#fnmatch.fnmatch) to match things passed in to things listed in the `skip` list. The match for `cextern` (and other `extern`) would occur at the directory level when `codespell` is parsing over the directory structure meaning `*extern` would indeed match and then it would not recurse to subdirectories. However when passing a file inside `cextern` directly, like `pre-commit` does, this results in the filename (which includes the file path) not matching `*extern` because we have not indicated a wildcard for suffix matching. This change adds a wildcard for the suffix which should cause the desired effect of files inside `cextern` to be properly skipped.

File: .pre-commit-config.yaml
Comment (WilliamJamieson) [inline .pre-commit-config.yaml:72]:
```suggestion
```
Can be removed if one makes the `pyproject.toml` change below.

**Review by WilliamJamieson — CHANGES_REQUESTED** [review]

File: pyproject.toml
Comment (namurphy) [reply pyproject.toml:270]: Thank you for digging into this!  Let's give it a try.

We have a saying over at PlasmaPy that:

> Sphinx rabbit holes often have dragons in them.

Maybe this saying can be applied to pre-commit as well? 👀 🐉 🕳️ 🐰

**Review by WilliamJamieson — APPROVED** [review]

**Comment by nstarman** [comment]
Failures are unrelated. Thanks @namurphy!

**Comment by pllim** [comment]
CI on main should be green now. Please rebase to make sure it stays green here. Thanks!

**Review by pllim — APPROVED** [review]
Thanks!

**Pull request merged**
